src/deparse.c

```diff
@@ -0,0 +1,260 @@
+#include "deparse.h"
+
+/**
+ * Deparse a table into a set of SQL commands that can be used to recreate it.
+ * Together with column definiton it deparses constraints, indexes, triggers and rules as well.
+ * There are some table types that are not supported: temporary, partitioned,
+ * foreign, inherited and a table that uses options. Row security is also not supported.
+ */
+
+TS_FUNCTION_INFO_V1(ts_deparse_table);
+
+Datum
+ts_deparse_table(PG_FUNCTION_ARGS)
+{
+	Oid			relid = PG_GETARG_OID(0);
+
+	PG_RETURN_TEXT_P(cstring_to_text(deparse_table(relid)));
+}
+
+static void
+deparse_columns (StringInfo stmt, Relation rel)
+{
+	TupleConstr *constraints = rel->rd_att->constr;
+
+	for (int i = 0; i < rel->rd_att->natts; i++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(rel->rd_att, i);
+
+		if (attr->attisdropped)
+			continue;
+
+		appendStringInfo(stmt, "\"%s\" %s",
+						 NameStr(attr->attname),
+						 format_type_with_typemod_qualified(attr->atttypid, attr->atttypmod));
+
+		if (attr->attnotnull)
+			appendStringInfoString(stmt, " NOT NULL");
+
+		if (OidIsValid(attr->attcollation))
+			appendStringInfo(stmt, " COLLATE \"%s\"",
+							 get_collation_name(attr->attcollation));
+
+		if (attr->atthasdef)
+		{
+			for (int i = 0; i < constraints->num_defval; i++)
+			{
+				AttrDefault attr_def = constraints->defval[i];
+
+				if (attr->attnum == attr_def.adnum)
+				{
+					char	   *attr_default = TextDatumGetCString(DirectFunctionCall2(pg_get_expr,
+																					   CStringGetTextDatum(attr_def.adbin),
+																					   ObjectIdGetDatum(rel->rd_id)));
+
+					appendStringInfo(stmt, " DEFAULT %s", attr_default);
+					break;
+				}
+			}
+		}
+
+		for (int i = 1; i < attr->attndims; i++)
+		{
+			appendStringInfoString(stmt, "[]");
+		}
+
+		if (i != (rel->rd_att->natts - 1))
+			appendStringInfoString(stmt, ", ");
+	}
+	appendStringInfoString(stmt, ") ");
+}
+
+static void
+deparse_constraints(StringInfo stmt, Oid relid, List **constraint_indexes)
+{
+	ScanKeyData skey;
+	Relation	constraint_rel;
+	SysScanDesc scan;
+	HeapTuple	htup;
+	char	   *constraint_stmt;
+
+	ScanKeyInit(&skey,
+				Anum_pg_constraint_conrelid,
+				BTEqualStrategyNumber, F_OIDEQ, relid);
+
+	constraint_rel = heap_open(ConstraintRelationId, AccessShareLock);
+	scan = systable_beginscan(constraint_rel, ConstraintRelidTypidNameIndexId, true,
+							  NULL, 1, &skey);
+
+	while (HeapTupleIsValid(htup = systable_getnext(scan)))
+	{
+		Form_pg_constraint pg_constraint = (Form_pg_constraint) GETSTRUCT(htup);
+
+		constraint_stmt = pg_get_constraintdef_command(HeapTupleGetOid(htup));
+		if (pg_constraint->conindid != 0)
+			*constraint_indexes = lappend_oid(*constraint_indexes, pg_constraint->conindid);
+		appendStringInfo(stmt, "%s;", constraint_stmt);
+	}
+
+	systable_endscan(scan);
+	heap_close(constraint_rel, AccessShareLock);
+}
+
+static void
+deparse_indexes(StringInfo stmt, Relation rel, List *exclude_indexes)
+{
+	HeapTuple	index_tuple;
```

Comment:
Move the declarations here to local context (within the loop), except for `ListCell`